File: 0-Linked-Lists/206-Reverse-Linked-Lists/206-Reverse-Linked-Lists-Iterative.java
/**
 * Problem 206: Reverse Linked List
 * Prompt: Reverse a singly linked list.
 * Date: 06/03/2020
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null)
            return null;
        ListNode endMark = head;
        while (endMark.next != null)
            endMark = endMark.next;
        ListNode curRead = head;
        while (curRead != endMark)
        {
            ListNode temp = new ListNode(curRead.val);
            temp.next = endMark.next;
            endMark.next = temp;
            curRead = curRead.next;
        }
        return endMark;
    }
}

/**
 * Notes: Iterative solution. 
 * Find the end of the linked list. Make a copy of the node each time 
 * and then insert it right after the end marker. Finally return
 * the end marker. 
 * */

File: 0-Linked-Lists/206-Reverse-Linked-Lists/206-Reverse-Linked-Lists-Iterative.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode * curNew = head;
        while (curNew->next != nullptr)
            curNew = curNew->next;
        ListNode * cur = head;
        while (cur != curNew)
        {
            ListNode * temp = new ListNode(cur->val);
            temp->next = curNew->next;
            curNew->next = temp;
            cur = cur->next;
        }
        return curNew;
    }
};

/**
 * Notes: Iterative solution. 
 * Find the end of the linked list. Make a copy of the node each time 
 * and then insert it right after the end marker. Finally return
 * the end marker. 
 * */